neps/nep-0508.md

```diff
@@ -0,0 +1,264 @@
+
+* 0 -  `:b`
+* 1 - `b:d`
+* 2 - `d:`
+
+Now if we add a fixed shard `f` to the same shard layout, then any we'll have 4 shards but neither is contiguous. Accounts such as `aaa.f`, `ccc.f`, `eee.f` that would otherwise belong to shards 0, 1 and 2 respectively are now all assigned to the fixed shard and create holes in the shard account ranges.  
+
+It's also worth noting that there is no benefit to having accounts colocated in the same shard. Any transaction or receipts is treated the same way regardless of crossing shard boundary. 
+
+This was implemented ahead of this NEP and the fixed shards feature was **removed**. 
+
+### Garbage collection
+
+In epoch T+2 once resharding is completed, we can delete the trie state and the flat state related to the parent shard. In practice, this is handled as part of the garbage collection code. While garbage collecting the last block of epoch T+1, we go ahead and clear all the data associated with the parent shard from the trie cache, flat storage, and RocksDB state associated with trie state and flat storage.
+
+### Transaction pool
+
+The transaction pool is sharded i.e. it groups transactions by the shard where each transaction should be converted to a receipt. The transaction pool was previously sharded by the ShardId. Unfortunately ShardId is insufficient to correctly identify a shard across a resharding event as ShardIds change domain. The transaction pool was migrated to group transactions by ShardUId instead, and a transaction pool resharding was implemented to reassign transaction from parent shard to children shards right before the new shard layout takes effect. The ShardUId contains the version of the shard layout which allows differentiating between shards in different shard layouts. 
+
+This was implemented ahead of this NEP and the transaction pool is now fully **migrated** to ShardUId. 
+
+## Alternatives
+
+### Why is this design the best in the space of possible designs?
+
+This design is the simplest, most robust and safe while meeting all of the requirements. 
+
+### What other designs have been considered and what is the rationale for not choosing them?
+
+#### Alternative implementations
+
+* Splitting the trie by iterating over the boundaries between children shards for each trie record type. This implementation has the potential to be faster but it is more complex and it would take longer to implement. We opted in for the much simpler one using flat storage given it is already quite performant.
+* Changing the trie structure to have the account id first and type of record later. This change would allow for much faster resharding by only iterating over the nodes on the boundary. This approach has two major drawbacks without providing too many benefits over the previous approach of splitting by each trie record type.
+  1) It would require a massive migration of trie.
+  2) We would need to maintain the old and the new trie structure forever.
```

Comment:
> Changing the trie structure to have the account id first and type of record later.

What is the `type of record`?

> We would need to maintain the old and the new trie structure forever.

Out of curiosity, why would we need to maintain the old structure?

Reply:
> What is the `type of record`?

The trie is an implementation of a key-value storage that supports multiple different value types. A single account may own multiple key-values, or records in the trie. For example an account with a deployed contract may have one record for storing the Account data, one record for storing the ContractCode and multiple records for storing ContractData. 

Here is the list of all possible types:
https://github.com/near/nearcore/blob/abf6e99bad479a216a7f6e8a465a3eaf0523659e/core/primitives/src/trie_key.rs#L16

> Out of curiosity, why would we need to maintain the old structure? 

The old structure was commited into the blockchain in the form of the state root merkle hash of the trie. Changing the trie structure would change the merkle hash. In order to be able to deal with historical data we would need to keep the old implementation in the code. 

Arguably this may only be needed for archival nodes and RPC nodes would need to maintain both only for a transitional period. In practice we would likely be forced to keep it foreve unless willing to fork the neard implementation for archival nodes.

Reply:
By committed to the blockchain I mean that the state roots were included in the headers that were then signed by producers and approved by validators.